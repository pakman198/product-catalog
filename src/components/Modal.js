import React from 'react';
import ReactDOM from 'react-dom';

import Carousel from './Carousel';

class Modal extends React.Component {

  renderModal() {
    const { title, images, onDismiss } = this.props;

    return (
      <div 
        className="ui dimmer modals visible active"
        onClick={() => onDismiss()}
      >
        <div 
          className="ui modal tiny visible active"
          onClick={e => e.stopPropagation()}
        >
          <div className="header">{title}</div>
          <div className="content">
            <Carousel images={images} />
          </div>
        </div>
      </div>
    );
  }

  componentWillUnmount() {
    document.querySelector('#modal').classList.remove('visible');
  }

  render() {
    const node = document.querySelector('#modal')
    node.classList.add('visible');
    const modal = this.renderModal();

    return ReactDOM.createPortal(modal, node);
  }
}

export default Modal;