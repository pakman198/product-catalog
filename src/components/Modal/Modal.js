import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';

import Carousel from './Carousel/Carousel';
import { ModalContext } from '../../contexts/ModalContext';

const Modal = () => {

  const renderModal = () => {
    return (
      <ModalContext.Consumer>
        {
          ({ currentItem, toggleModal }) => (
            <div 
              className="ui dimmer modals visible active"
              onClick={() => toggleModal()}
            >
              <div 
                className="ui modal tiny visible active"
                onClick={e => e.stopPropagation()}
              >
                <div className="header">{currentItem.name}</div>
                <div className="content">
                  <Carousel images={currentItem.images} />
                </div>
              </div>
            </div>
          )
        }
      </ModalContext.Consumer>
    );
  }

  useEffect(() => {
    return () => {
      document.querySelector('#modal').classList.remove('visible');
    }
  });

  const node = document.querySelector('#modal')
  node.classList.add('visible');
  const modal = renderModal();

  return ReactDOM.createPortal(modal, node);
}

export default Modal;