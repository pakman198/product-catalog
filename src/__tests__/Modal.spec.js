import React from 'react';
import { 
  render, 
  fireEvent, 
  waitForElement,
} from '@testing-library/react'
import '@testing-library/jest-dom';

import Modal from '../components/Modal';

describe('Modal.js', () => {

  const props = {
    onDismiss: jest.fn(),
    title: "Modal Header",
    images: [
      {
        "size": "m",
        "meta": "",
        "alt": "",
        "rel": "althero",
        "width": 363,
        "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0011/tahoe-solid-wood-dining-table-natural-oak-m.jpg",
        "height": 363
      },
      {
        "size": "m",
        "meta": "",
        "alt": "",
        "rel": "althero",
        "width": 363,
        "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0012/tahoe-solid-wood-dining-table-natural-oak-2-m.jpg",
        "height": 363
      },
    ]
  }
  
  const modalRoot = document.createElement('div')
  modalRoot.setAttribute('id', 'modal')
  
  test('Modal is displayed', () => {
    const { asFragment } = render(<Modal {...props} />, {
      container: document.body.appendChild(modalRoot),
    });
    
    expect(asFragment()).toMatchSnapshot()
  });

  test('Click on modal does nothing', async () => {
    const { container, getByText } = render(<Modal {...props} />, {
      container: document.body.appendChild(modalRoot),
    });

    let header = await waitForElement(() => container.querySelector('.header')); 

    fireEvent.click(header);
    header = await waitForElement(() => getByText('Modal Header'))

    expect(header.classList).toContain('header');
    expect(props.onDismiss).toHaveBeenCalledTimes(0);
  });

  test('Click outside modal closes it', async() => {
    const { container } = render(<Modal {...props} />, {
      container: document.body.appendChild(modalRoot),
    });
    let dimmer = await waitForElement(() => container.querySelector('.dimmer'));

    fireEvent.click(dimmer)
    
    expect(props.onDismiss).toHaveBeenCalledTimes(1);
  });

});