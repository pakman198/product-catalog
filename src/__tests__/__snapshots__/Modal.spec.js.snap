// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.js Modal is displayed 1`] = `
<DocumentFragment>
  <div
    class="ui dimmer modals visible active"
  >
    <div
      class="ui modal tiny visible active"
    >
      <div
        class="header"
      >
        Modal Header
      </div>
      <div
        class="content"
      >
        <div
          class="slider"
        >
          <div
            class="slider-wrapper"
            style="transform: translateX(0px);"
          >
            <div
              class="slide"
              style="background-image: url(https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0011/tahoe-solid-wood-dining-table-natural-oak-m.jpg);"
            />
            <div
              class="slide"
              style="background-image: url(https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0012/tahoe-solid-wood-dining-table-natural-oak-2-m.jpg);"
            />
          </div>
          <button
            aria-label="prev-slide"
            class="ui inverted secondary circular angle left icon button left-arrow"
          >
            <i
              class="angle left icon"
            />
          </button>
          <button
            aria-label="next-slide"
            class="ui inverted secondary circular angle right icon button right-arrow"
          >
            <i
              class="angle right icon"
            />
          </button>
          <span
            data-testid="current-index"
            style="display: none;"
          >
            0
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
